class.jetpack-gutenberg.php

```diff
@@ -10,143 +10,152 @@
+	 *
+	 * @param string $slug Slug of the block.
+	 * @param array  $args Arguments that are passed into register_block_type().
 	 */
-	public static function init() {
-		if ( ! self::is_gutenberg_available() ) {
-			return;
-		}
-
-		if ( ! self::should_load() ) {
-			return;
+	public static function register_block( $slug, $args ) {
+		$prefixed_extensions = array_map( array( __CLASS__, 'prepend_block_prefix' ), self::$extensions );
+
+		// Register the block if it's whitelisted, or if it's a child block, and one of its parents is whitelisted.
+		if ( in_array( $slug, self::$extensions ) || ( isset( $args['parent'] ) && self::share_items( $args['parent'], $prefixed_extensions ) ) ) {
+			register_block_type( 'jetpack/' . $slug, $args );
+		} else if ( ! isset( $args['parent'] ) ) { // Don't set availability information for child blocks -- we infer it from their parents
+			self::set_extension_unavailability_reason( $slug, 'not_whitelisted' );
 		}
+	}
 
-		if ( Jetpack_Constants::is_true( 'REST_API_REQUEST' ) ) {
-			// We defer the loading of the blocks until we have a better scope in reset requests.
-			add_filter( 'rest_request_before_callbacks', array( __CLASS__, 'load' ), 10, 3 );
-			return;
+	/**
+	 * Register a plugin
+	 *
+	 * If the plugin isn't whitelisted, set its unavailability reason instead.
+	 * 
+	 * @param string $slug Slug of the plugin.
+	 */
+	public static function register_plugin( $slug ) {
+		if ( in_array( $slug, self::$extensions ) ) {
+			self::$registered_plugins[] = 'jetpack-' . $slug;
+		} else {
+			self::set_extension_unavailability_reason( $slug, 'not_whitelisted' );
 		}
-		self::load();
 	}
 
 	/**
-	 * @deprecated
+	 * Set the reason why an extension (block or plugin) is unavailable
+	 *
+	 * @param string $slug Slug of the extension.
+	 * @param string $reason A string representation of why the extension is unavailable
 	 */
-	static function load_blocks() {
-		self::init();
+	public static function set_extension_unavailability_reason( $slug, $reason ) {
+		self::$availability[ $slug ] = $reason;
 	}
 
 	/**
-	 * Add a block to the list of blocks to be registered.
+	 * Set up a whitelist of allowed block editor extensions
 	 *
-	 * @param string $slug Slug of the block.
-	 * @param array  $args Arguments that are passed into the register_block_type.
-	 * @param array $availability array containing if a block is available and the reason when it is not.
-	 *
-	 * @deprecated
+	 * @return void
 	 */
-	static function add_block( $type, $args, $availability ) {
-		self::register( $type, $args, $availability );
-	}
+	public static function init() {
+		if ( ! self::is_gutenberg_available() ) {
+			return;
+		}
 
-	static function load( $response = null, $handler = null, $request = null ) {
-		$is_availability_endpoint_beta = ! is_null( $request ) && $request->get_param( 'beta' ) && wp_endswith( $request->get_route(), 'gutenberg/available-extensions' );
```

Comment:
We seem to unconditionally be removing this logic. Is that intentional? Are we sure we're not relying on it somewhere?

Reply:
The whole idea behind the flag was to add the ability for any API client to be able to include the beta blocks weather the user is proxied or not. Removing the argument make that impossible.